edgelesssys/constellation
# upgrade: support Kubernetes components

### Proposed change(s)

- Rename `NodeImage` and friends to `NodeVersion` 
- Add Kubernetes components to `NodeVersion`
- Let the JoinService look up the current components ConfigMap in the `NodeVersion` custom resource instead of having a separate `k8s-version` ConfigMap

Discussion: Should we announce this feature in the release notes already? 
My opinion: No, since it is part of the complete upgrade feature which needs to be fully implemented, documented, and tested.

How to test:
1. Build the JoinService and node operator from this branch and use it in `versions.go`
2. Create  a valid `k8s-components-sha256-ABC` ConfigMap (Download links and hashes need to match+work)
3. Reference new ConfigMap in `NodeVersion` named `node-versioin`
4. Observe the update process (see. `NodeVersion` CRD, `PendingNode` CRD, and `Noes`)

What works best for me is to downgrade the patch version of the Kubernetes components in the `versions.go`. Then create a Constellation and upgrade to the latest patch version. This version should be visible via `kubectl get nodes -A -o wide`.

### Checklist
